haskell/cabal.bzl

```diff
@@ -243,3 +243,225 @@ However, using a plain `haskell_library` sometimes leads to better
+    "ghci",
+    "haskeline",
+    "hpc",
+    "integer-gmp",
+    "libiserv",
+    "mtl",
+    "parsec",
+    "pretty",
+    "process",
+    "rts",
+    "stm",
+    "template-haskell",
+    "terminfo",
+    "text",
+    "time",
+    "transformers",
+    "unix",
+    "xhtml",
+]
+
+def _compute_dependency_graph(repository_ctx, versioned_packages, unversioned_packages):
+    """Given a list of root packages, compute a dependency graph.
+
+    Returns:
+      dependencies: adjacency list of packages, represented as a dictionary.
+      transitive_unpacked_sdists: directory names of unpacked source distributions.
+
+    """
+
+    if not versioned_packages and not unversioned_packages:
+        return ({}, [])
+
+    # Unpack all given packages, then compute the transitive closure
+    # and unpack anything in the transitive closure as well.
+    stack_cmd = repository_ctx.which("stack")
+    if not stack_cmd:
+        fail("Cannot find stack command in your PATH.")
+    stack = [stack_cmd, "--no-nix", "--skip-ghc-check", "--system-ghc"]
+    if versioned_packages:
+        _execute_or_fail_loudly(repository_ctx, stack + ["unpack"] + versioned_packages)
+    stack = [stack_cmd, "--no-nix", "--skip-ghc-check", "--system-ghc", "--resolver", repository_ctx.attr.snapshot]
+    if unversioned_packages:
+        _execute_or_fail_loudly(repository_ctx, stack + ["unpack"] + unversioned_packages)
+    exec_result = _execute_or_fail_loudly(repository_ctx, ["ls"])
+    unpacked_sdists = exec_result.stdout.splitlines()
+    stack_yaml_content = struct(resolver = "none", packages = unpacked_sdists).to_json()
+    repository_ctx.file("stack.yaml", content = stack_yaml_content, executable = False)
+    exec_result = _execute_or_fail_loudly(
+        repository_ctx,
+        stack + ["ls", "dependencies", "--separator=-"],
+    )
+    transitive_unpacked_sdists = [
+        unpacked_sdist
+        for unpacked_sdist in exec_result.stdout.splitlines()
+        if _chop_version(unpacked_sdist) not in _CORE_PACKAGES
+    ]
+    _execute_or_fail_loudly(
+        repository_ctx,
+        stack + ["unpack"] + [
+            unpacked_sdist
+            for unpacked_sdist in transitive_unpacked_sdists
+            if unpacked_sdist not in unpacked_sdists
+        ],
+    )
+    stack_yaml_content = struct(resolver = "none", packages = transitive_unpacked_sdists).to_json()
+    repository_ctx.file("stack.yaml", stack_yaml_content, executable = False)
+
+    # Compute dependency graph.
+    all_packages = [_chop_version(dir) for dir in transitive_unpacked_sdists + _CORE_PACKAGES]
+    exec_result = _execute_or_fail_loudly(
+        repository_ctx,
+        stack + ["dot", "--external"],
+    )
+    dependencies = {k: [] for k in all_packages}
+    for line in exec_result.stdout.splitlines():
+        tokens = [w.strip('";') for w in line.split(" ")]
+
+        # All lines of the form `"foo" -> "bar";` declare edges of the
+        # dependency graph in the Graphviz format.
```

Comment:
It would be great if stack could spit this out in a more robust and easier parseable format.

Reply:
Agreed. I'll file a ticket upstream.